src/python/espressomd/reaction_ensemble.py

```diff
@@ -0,0 +1,567 @@
+
+        if "product_coefficients" in kwargs:
+            if kwargs["product_coefficients"][0] != 1 or kwargs["product_coefficients"][1] != 1:
+                raise ValueError(err_msg)
+            else:
+                warn_user = True
+        else:
+            kwargs["product_coefficients"] = [1, 1]
+
+        if warn_user:
+            warnings.warn(warn_msg, FutureWarning)
+
+        if(len(kwargs["product_types"]) != 2 or len(kwargs["reactant_types"]) != 1):
+            raise ValueError(
+                "The constant pH method is only implemented for reactions "
+                "with two product types and one adduct type.")
+
+        super().add_reaction(*args, **kwargs)
+
+
+@script_interface_register
+class WidomInsertion(ReactionAlgorithm):
+    """
+    This class implements the Widom insertion method in the canonical ensemble
+    for homogeneous systems, where the excess chemical potential is not
+    depending on the location.
+
+    """
+
+    _so_name = "ReactionMethods::WidomInsertion"
+    _so_creation_policy = "LOCAL"
+
+    def required_keys(self):
+        return {"kT", "seed"}
+
+    def valid_keys(self):
+        return {"kT", "seed"}
+
+    def add_reaction(self, **kwargs):
+        kwargs['gamma'] = 1.
+        super().add_reaction(**kwargs)
+
+    def calculate_particle_insertion_potential_energy(self, **kwargs):
+        """
+        Measures the potential energy when particles are inserted in the
+        system following the reaction  provided ``reaction_id``. Please
+        define the insertion moves first by calling the method
+        :meth:`~ReactionAlgorithm.add_reaction` (with only product types
+        specified).
+
+        Note that although this function does not provide directly
+        the chemical potential, it can be used to calculate it.
+        For an example of such an application please check
+        :file:`/samples/widom_insertion.py`.
+        """
+        # make inverse widom scheme (deletion of particles) inaccessible.
+        # The deletion reactions are the odd reaction_ids
+
+        return self.call_method(
+            "calculate_particle_insertion_potential_energy", reaction_id=kwargs.pop("reaction_id"))
+
+    def calculate_excess_chemical_potential(
+            self, **kwargs):
+        """
+        Given a set of samples of the particle insertion potential energy,
+        calculates the excess chemical potential and its statistical error.
+
+        Parameters
+        ----------
+        particle_insertion_potential_energy_samples : array_like of :obj:`float`
+            Samples of the particle insertion potential energy.
+        N_blocks : :obj:`int`, optional
+            Number of bins for binning analysis.
+
+        Returns
+        -------
+        mean : :obj:`float`
+            Mean excess chemical potential.
+        error : :obj:`float`
+            Standard error of the mean.
+        """
+
+        def do_block_analysis(samples, N_blocks=16):
```

Comment:
```suggestion
        def do_block_analysis(samples, N_blocks):
```
This default-value was never necessary in this code, so we can just remove it

Reply:
Done.